include/zephyr/storage/storage_area/storage_area.h

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright (c) 2024 Laczen
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public API for storage area subsystem
+ *
+ * The storage area API is a subsystem that creates a unified method to work
+ * with flash, eeprom, ram, disks, files, ... for storage. A storage area is
+ * an area that has a number of constant sized erase blocks, and has constant
+ * write block size. The storage area does not necessarily inherit the
+ * limitations of the underlying storage device but rather defines a method of
+ * how the underlying storage device will be used (however it does not remove
+ * any limitations of the underlying storage device). A storage area can be
+ * defined with wrong properties, optionally the definition will be checked
+ * by defining the Kconfig option `CONFIG_STORAGE_AREA_VERIFY` that is by
+ * default enabled for `DEBUG` builds.
+ *
+ *
+ * There following methods are exposed:
+ *   storage_area_read(),	** read data **
+ *   storage_area_readv(),	** read data vector **
+ *   storage_area_write(),	** write data **
+ *   storage_area_writev(),	** write data vector **
+ *   storage_area_erase(),	** erase (in erase block addressing) **
+ *   storage_area_ioctl()	** used for e.g. getting xip addresses **
+ *
+ * The subsystem is easy extendable to create custom (virtual) storage areas
+ * that consist of e.g. a combination of flash and ram, an encrypted storage
+ * area, ...
+ *
+ * A storage area is defined e.g. for flash:
+ * struct storage_area_flash fa =
+ *	flash_storage_area(dev, start, xip_address, write_size, erase_size,
+ *			   size, properties);
+ * struct storage_area *area = &fa.area;
+ *
+ * For other storage devices (eeprom, ram, disk, ...) similar macros are
+ * defines, but they might differ slightly.
+ *
+ * The write_size, erase_size, ... are declarations of how the storage_area
+ * will be used The write_size is limited to a power of 2, erase_size should
+ * be a multiple of write_size and size should be a multiple of erase_size.
+ * The macro definitions xxxxx_storage_area(...) checks these conditions but
+ * always succeeds. Trying to use a badly sized storage area will result in
+ * failure of any of the exposed methods.
+ *
+ */
+
+#ifndef ZEPHYR_INCLUDE_STORAGE_STORAGE_AREA_H_
+#define ZEPHYR_INCLUDE_STORAGE_STORAGE_AREA_H_
+
+#include <stdint.h>
+#include <stddef.h>
+#include <sys/types.h>
+#include <zephyr/sys/util.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Storage_area interface
+ * @defgroup storage_area_interface Storage_area interface
+ * @ingroup storage_apis
+ * @{
+ */
+
+/* Allow the sa_off_t type to be redefined if needed */
+typedef off_t sa_off_t;
```

Comment:
Please use an appropriate Kconfig option here and define `sa_off_t` as e.g. a `uint64_t` or `uint32_t`.

```suggestion
#ifdef CONFIG_STORAGE_AREA_OFFSET_64
typedef uint64_t sa_off_t;
#else
typedef uint32_t sa_off_t;
#endif
```